### ISSUE TITLE
Ski or Die timing issue

### ISSUE THREAD

**Issue by jschwartzenberg** [dosemu2/dosemu2#157]
I'm not sure whether this is a dosemu bug, but I have never seen this before and can't reproduce it in any other environment. (One example, dosbox without CPU limit does work correctly.)

When playing the game Ski or Die, there are issues with movement which is especially noticeable when picking 'Downhill Blitz', but also makes the other modes much more difficult.
When moving left or right for instance, the character moves way too fast. Also the animation when falling is too fast.

This does not seem to be a recent regression, the 1.5.0 fork point version has the same issue. More information about the game:
http://www.abandonia.com/en/games/22/Ski+or+Die.html

**Comment by stsp** [comment]
Yes, completely unplayable. :(
Have you tried that on a bare-metal DOS?

**Comment by stsp** [comment]
And no music too. :(
Since this is my favourite game of the
deepest past, I'd really like it to be supported. :)

**Comment by jschwartzenberg** [comment]
The music works fine here in dosemu when I run it with the adlib argument. I haven't been able to try it in bare-metal DOS, the newest PC I have with a bare-metal DOS installation is a Pentium 3 700MHz, but I've only tested it on faster systems with dosemu.

The NES/Dendy version of this game is nice too btw :)

**Comment by stsp** [comment]
Yes, seems to work with adlib.
But the pc-speaker music there is quite cool too. :)

**Comment by stsp** [comment]
The CPU speed does not seem to affect this.
Timer speed does.
So no need to test on fast machines.

**Comment by stsp** [comment]
Well I guess our timer code is so much piece
of shit that I can't follow it. :) Needs to be replaced
with something from qemu.

**Comment by stsp** [comment]
In fact I already created the pic_cleanups2
branch where I aimed for a complete rewrite,
but its a bit stalled.

**Comment by stsp** [comment]
Interesting.
When you just enter
'Downhill Blitz', the character
has some animation, waiting
for you to start the race.
This animation is very fast,
and it is dependent on a CPU
speed. It is also very fast
under dosbox.

But as soon as you start
the race, the speed depends
on a timer. And it remains too
fast under dosemu but drops
to normal under dosbox.

So what can be the problem?
It uses PIT mode 3, which I
think is similar to mode 2 wrt
interrupt rate. Some latching
of pit2 is happening but the
speed doesn't depend on a
latched values (I can make it
always latch the same number).
Weird modes (like bcd) are
not seems to be used.
So the only guess I can make
is that it does some calibration.
But its not the case too, after
I checked the pit counter value
on dosbox, and its the same
as on dosemu.

So in short: no idea what makes
it faster than it should.

**Comment by stsp** [comment]
Could you please check under qemu?

**Comment by jschwartzenberg** [comment]
Yeah I guess so. I should look into making that easier for myself :)

Btw. this has nothing to do with `$_timer_tweaks=(on)`? That fix is hilarious btw. not enabling it is essentially cheating, at least for Tetris Classic :) :)

**Comment by jschwartzenberg** [comment]
And wow it's been so many years already, I had forgotten about this issue!!

**Comment by stsp** [comment]
> Btw. this has nothing to do with $_timer_tweaks=(on)?

Can't find any relation
with that, either. Ski could
disable timer interrupts for
too long to lose some, which
I could detect with
`$_timer_tweaks=(on)`
but that's not the case.

> That fix is hilarious btw. not enabling it is essentially cheating, at least for Tetris Classic :) :)

Unfortunately this is not a
fix. The bug is in the game,
and I actually have binary
patches for all affected games.
Which is why it is not enabled
by default. I don't know if it
will break some other apps.

> And wow it's been so many years already, I had forgotten about this issue!!

Things with dosemu2 development
are not as fast as in this Downhill
Blitz game, unfortunately. :)
I've just got to writing the timer
subsystems this year, and currently
am writing the event-based timers
for #1639. As it is mostly complete,
I am checking other problems, like
this one.
Its quite amazing how dosemu1
got around everything w/o writing
seemingly any code at all (entire
timers stack is a small hack in a
pic code), but this means that
dosemu2 should mostly be written
from scratches.

**Comment by jschwartzenberg** [comment]
> The bug is in the game,
> and I actually have binary
> patches for all affected games.

Quite a few of these patches are also in the the tweaks document in the repo. Have you considered providing those there too? Maybe it would be good to keep a separate repository with binary patches for these kinds of bugfixes? They might help on other non-DOSEMU environments too.

**Comment by stsp** [comment]
I think I removed them from docs when
added an option. I am not sure if many
people are going to apply these patches.
Since dosemu2 is about implementing
things in an unusual way, I think we need
to just upload the patched binaries
somwhere first. And then think up some
kind of packaging them and automating
the replacements.

**Comment by jschwartzenberg** [comment]
I was just searching a bit on doing this from Python and came across this:
https://stackoverflow.com/questions/14643538/modifying-binary-file-with-python

It seems it wouldn't be too hard to write a Python script and similarly let it be called from within dosemu through a batch file. This would let end users quickly patch known applications from within dosemu.

**Comment by stsp** [comment]
Or do other emulators already have
something like that?

**Comment by stsp** [comment]
I suppose we can just replace the binary
on start, so that the user only have to
install the package. His original binary
will remain, but the patched one will
instead be starting.
The loadfix problem should be the biggest
user of that.

**Comment by jschwartzenberg** [comment]
That would be really neat, although I could still imagine that there would also be other environments that could benefit from these kinds of patches. Maybe problems are less noticeable in non-dosemu environments, but are still there and these patches could help more people?

**Comment by stsp** [comment]
The problem with patches is
that many progs needs to be
unpacked before applying a
patch. This is especially the
case with "packed file corrupt"
problem where buggy packer
was used.

In general, how is the textual
patch better than the patched
binary that we can provide and
even package?

**Comment by stsp** [comment]
Ppl can recreate the patch from
the binary, if they want.

**Comment by stsp** [comment]
Or alternatively can you find the host
unpackers and eg unpack ski?
Then indeed it would be better to apply
patches on a fly.

**Comment by jschwartzenberg** [comment]
In QEMU the timing seems good with the default accelerator, similar to DOSBox. With KVM it's quite weird, some things seem slower. It's playable still though.

**Comment by stsp** [comment]
Is the initial animation in
downhill blitz also good?

**Comment by jschwartzenberg** [comment]
It depends on whether KVM is used or not. I think with KVM it's slower than without.

**Comment by stsp** [comment]
Well.
It appears if you exit the
game after checking that
DB is too fast, and run it
again, you will eventually
get normal speed.
It requires about ~10
restarts of the game (no
need to restart entire dosemu)
to get the right speed.
Sometimes you even get
it from the first run, but that
is quite rare.

**Comment by stsp** [comment]
[skiega.gz](https://github.com/dosemu2/dosemu2/files/8884731/skiega.gz)
Here is the patched binary
that always gives the right
speed.

The game does some
calibration with pit 2, and
it only has 2 outcomes:
normal speed and double
speed. Perhaps they got
some buggy HW where
the timer is twice slower
than it should.

**Comment by stsp** [comment]
Added to Tweaks:
https://github.com/dosemu2/dosemu2/wiki/Tweaks#211-game-play-is-too-fast
